design-proposals/shadow-node.md

```diff
@@ -0,0 +1,210 @@
+```yaml
+apiVersion: <apiversion>
+kind: ShadowNode
+metadata:
+  name: <Equivalent to existing name>
+  # Labels and Annotations only
+spec: {} //Empty, allows further extension
+status: {} //Empty, allows further extension
+```
+
+### Solution option #2: Creating a REST/GRPC communication between virt-handler and virt-controller
+Pass the labels/annotations through a REST/GRPC channel to the virt-controller. REST option is preferred because the server-side is already implemented.
+
+Virt-handler (Server): Instead of directly patching labels/annotation to the node, Virt-handler will maintain a "cache" (NodeInfoMgr) where all the components will "set" their data.
+Virt handler can use its existing [web-service](https://github.com/kubevirt/kubevirt/blob/b102c56f0fcd52feff3ff7a6296737b8e8b99131/cmd/virt-handler/virt-handler.go#L556) to expose one more endpoint - GetNodeInfo. Once reached - this endpoint will return the node data.
+
+Virt-Controller (Client): Add a new web-service client infra that will:
+1. run as a go routine and poll the client for updates every one minute.This frequency is chosen because it is the highest of the frequencies (=heartbeat).
+2. ask every Virt-handler server separately for the NodeInfo (need to make sure it's possible to address them in a unique manner).
+3. filter non-kubevirt owned annotations/labels.
+4. patch the node if necessary.
+
+The heartbeat annotation and virt-controller's [Node-controller](https://github.com/kubevirt/kubevirt/blob/b102c56f0fcd52feff3ff7a6296737b8e8b99131/pkg/virt-controller/watch/node.go) can be removed in favor for the web-service timeout mechanism (if received timeout - set node as unreachable).
+
+Pros:
+- Avoid introducing a new CRD to the project.
+- Using existing web-server infra on the server side.
+- security:
+  - the web-service is secure using the cert-manager internal API.
+  - Virt-handler does not need to access kube-API at all, and cannot modify unrelated nodes or shadow nodes. Virt-controller does all the risky work
+- Scalability:
+  - Reduces #API requests (i.e. node patches) to a minimum.
+
+Cons:
+- More complex. It's easier and more appealing to use the k8s API, i.e. patching labels/annotations, than to implement and maintain GRPC/REST case logic internally.
+- Virt-handler will no longer be able to immediately set the node as unschedulable (in case of virt-handler SIGTERM). However, this use-case is pretty rare, and IMO should be reconsidered.
+
+## Solution option #3: Moving to third party entity
+The idea is to eliminate the reason(s) to patch to the node from virt-handler in the first place.
+
+1. Use tools like [Node Feature Discovery](https://kubernetes-sigs.github.io/node-feature-discovery/v0.15/get-started/introduction.html) or utilize virt-handler's [device-manager](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/) to publish node features (pending checking whether they don't have the CVE issue as well).
```

Comment:
I think it is worth to elaborate this idea a bit further. What do you intend to do with NFD and device plugins?

For example, setting `kubevirt.io/schedulable` may be replaced by tearing down the `devices.kubevirt.io/kvm` device.

Reply:
What do you mean by tearing down?

If you mean that we can monitor the `devices.kubevirt.io/kvm` device from virt-controller and set the `kubevirt.io/schedulable` in case it is removed/unhealthy - then I don't think that is supported. Meaning a device can only be set unhealthy manually, and kubelet does not support removing devices if the plugin connection is lost

Reply:
I'm asking to provide more detail on how you are going to use device manager and node feature discovery.

I raised an example, which you may be thinking about. I don't know what exposes the devices.kubevirt.io/kvm`, but that component may un-expose it if virt-handler is shut down, effectively making `kubevirt.io/schedulable` redundant.

Reply:
The thing is that virt-handler **is** the one running device-manager and exposes these devices to the node, so it can't be used to monitor itself..
I honestly don't think it can be used to replace the heartbeat, but again I'm open for suggestions. 

I will elaborate more on how device-manager plugin can be used to replace the node-labeller.

Reply:
DONE. PTAL